rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+is enabled. Like any normal Teleport client, the agent client will only prompt
+for PIN or touch when it isn't cached.
+
+#### `hardwareKeyAgentService` pseudo-code implementation
+
+```go
+// Pseudo-code
+
+// hardwareKeyRef uniquely identifies a hardware key
+type hardwareKeyRef struct {
+  serialNumber uint32
+  pivSlot      uint32
+}
+
+type hardwareKeyInfo struct {
+  publicKey            crypto.PublicKey
+  privateKeyPolicy     keys.PrivateKeyPolicy
+  attestationStatement keys.AttestationStatement
+}
+
+// hardwareKeyAgentService has two implementations:
+//  - direct implementation with piv-go, adapted slightly from our existing implementation
+//  - hardware key agent gRPC service implementation
+type hardwareKeyAgentService interface {
+  GetInfo(ref hardwareKeyRef) hardwareKeyInfo
+  Sign(ref hardwareKeyRef, digest []byte, opts SignerOpts) (signature []byte, err error)
+}
+
+// Implements [crypto.Signer] and [keys.HardwareSigner].
+type hardwareKeyAgentKey struct {
+  agent  keyAgentService
+  keyRef hardwareKeyRef
+  // hardwareKeyInfo is cached after the first call
+  hardwareKeyInfo hardwareKeyInfo
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Public() crypto.PublicKey {
+  return s.hardwareKeyInfo.publicKey
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+  return s.agent.Sign(as.KeyRef, digest, opts)
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetAttestationStatement() *AttestationStatement {
+  return s.hardwareKeyInfo.attestationStatement
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetPrivateKeyPolicy() PrivateKeyPolicy {
+  return s.hardwareKeyInfo.privateKeyPolicy
+}
+
+func (s *agentSigner) fetchHardwareKeyInfo() (hardwareKeyInfo, error) {
+  if s.hardwareKeyInfo == nil {
+    s.hardwareKeyInfo = s.agent.GetInfo(s.ref)
+  }
+  return s.hardwareKeyInfo
+}
+```
+
+### UX
+
+#### PIN caching
+
+The PIN caching portion of this RFD does not present any additional UX concerns.
+It is purely a UX benefitting change, removing the need for back-to-back PIN
+prompts.
+
+#### Agent clients
+
+A user can start the hardware key agent with Teleport Connect or `tsh`.
+
+##### Teleport Connect
+
+Teleport Connect is an ideal runner for the key agent because it:
+
+1. Is a long-lived process
+1. Provides a UI for touch/PIN prompts, and can foreground itself for these prompts
+
+By default, if `tshd` detects a hardware key, it will automatically run and
+serve the hardware key agent service. This way, the agent will be served
+regardless of Teleport Connect's [login state](#running-the-agent-before-login).
+
+For example, if a user primarily wants to use `tsh`, but get PIN caching and
+PIN prompts in Teleport Connect, they could just launch Teleport Connect without
+logging in. Teleport Connect would just foreground itself with hardware key
+prompts for the user as needed without adding additional overhead.
```

Comment:
This is a nice UX, but what happens if there's already another agent client running? Does connect detect this? Does it take the place of the existing agent or refuse to start?